runelite-client/src/main/java/net/runelite/client/ui/TitleToolbar.java

```diff
@@ -24,174 +25,73 @@
-			{
-				return 0;
-			}
-
-			@Override
-			public void invalidateLayout(Container target)
-			{
-			}
-
-			@Override
-			public void layoutContainer(Container parent)
-			{
-				int x = 0;
-				for (Component c : parent.getComponents())
-				{
-					x += ITEM_PADDING;
-					int height = c.getPreferredSize().height;
-					if (height > TITLEBAR_SIZE)
-					{
-						height = TITLEBAR_SIZE;
-					}
-					c.setBounds(x, (TITLEBAR_SIZE - height) / 2, TITLEBAR_SIZE, height);
-					x += TITLEBAR_SIZE;
-				}
-			}
-		});
+		this.eventBus = eventBus;
+	}
 
-		try
+	/**
+	 * Add navigation.
+	 *
+	 * @param button the button
+	 */
+	public void addNavigation(final NavigationButton button)
+	{
+		if (buttons.contains(button))
 		{
-			BufferedImage discordIcon;
-			BufferedImage invertedIcon;
-			synchronized (ImageIO.class)
-			{
-				discordIcon = ImageIO.read(ClientUI.class.getResourceAsStream("discord.png"));
-				invertedIcon = ImageIO.read(ClientUI.class.getResourceAsStream("discord_inverted.png"));
-			}
-
-			JButton discordButton = new JButton();
-			discordButton.setToolTipText("Join Discord");
-			discordButton.addMouseListener(new MouseAdapter()
-			{
-				@Override
-				public void mouseClicked(MouseEvent e)
-				{
-					super.mouseClicked(e);
-					try
-					{
-						Desktop.getDesktop().browse(new URL(properties.getDiscordInvite()).toURI());
-					}
-					catch (IOException | URISyntaxException ex)
-					{
-						log.warn("error opening browser", ex);
-					}
-				}
-			});
-
-			addButton(discordButton, discordIcon, invertedIcon);
+			return;
 		}
-		catch (IOException ex)
+
+		if (buttons.add(button))
 		{
-			log.warn("unable to load discord button", ex);
+			int index = buttons.headSet(button).size();
+			eventBus.post(new TitleToolbarButtonAdded(button, index));
 		}
 	}
 
-	public void addButton(JButton button, Image iconImage, Image invertedIconImage)
+	/**
+	 * Remove navigation.
+	 *
+	 * @param button the button
+	 */
+	public void removeNavigation(final NavigationButton button)
 	{
-		final int iconSize = TITLEBAR_SIZE - 6;
-		ImageIcon icon = new ImageIcon(iconImage.getScaledInstance(iconSize, iconSize, Image.SCALE_SMOOTH));
-		ImageIcon invertedIcon;
-		if (invertedIconImage == null)
-		{
-			invertedIcon = icon;
-		}
-		else
+		int index = buttons.headSet(button).size();
+
+		if (buttons.remove(button))
 		{
-			invertedIcon = new ImageIcon(invertedIconImage.getScaledInstance(iconSize, iconSize, Image.SCALE_SMOOTH));
+			eventBus.post(new TitleToolbarButtonRemoved(button, index));
 		}
-
-		button.setIcon(icon);
-		button.setRolloverIcon(invertedIcon);
-		button.putClientProperty(SubstanceSynapse.FLAT_LOOK, Boolean.TRUE);
-		button.setFocusable(false);
-
-		add(button);
-		revalidate();
-		repaint();
 	}
 
-	@Override
-	public void remove(Component c)
+	/**
+	 * Refresh all buttons
+	 */
+	public void refresh()
 	{
-		super.remove(c);
-		revalidate();
-		repaint();
+		buttons.forEach(button ->
```

Comment:
This should be an iterator, as it is this is O(n log n) where it should be O(n). Also, I don't understand what this is for?

Reply:
This is for functionality you asked in InfoPanel, it sends events for all loaded buttons again, so InfoPanel can catch them and add the buttons.

Reply:
Updated to use iterator and manually make index